fungible_conversion_proxy/src/lib.rs

```diff
@@ -0,0 +1,659 @@
+        args: PayerSuppliedArgs,
+        payer: AccountId,
+        deposit: U128,
+    ) -> Promise {
+        near_sdk::assert_self();
+
+        // Parse fungible token metadata from promise result
+        let ft_metadata = match env::promise_result(0) {
+            PromiseResult::NotReady => unreachable!(),
+            PromiseResult::Successful(value) => {
+                match serde_json::from_slice::<FungibleTokenMetadata>(&value) {
+                    Ok(value) => value,
+                    Err(_e) => panic!("ERR_INVALID_FT_METADATA_RESPONSE"),
+                }
+            }
+            PromiseResult::Failed => panic!("ERR_FAILED_FT_METADATA_FETCH"),
+        };
+
+        let get_rate = fpo_contract::get_entry(
+            ft_metadata.symbol + "/" + &args.currency,
+            self.provider_account_id.clone(),
+            &self.oracle_account_id,
+            NO_DEPOSIT,
+            BASIC_GAS,
+        );
+        let process_request_payment = ext_self::rate_callback(
+            args,
+            payer,
+            deposit,
+            ft_metadata.decimals,
+            &env::current_account_id(),
+            env::attached_deposit(),
+            BASIC_GAS * 8,
+        );
+        get_rate.then(process_request_payment)
+    }
+
+    #[private]
+    #[payable]
+    pub fn rate_callback(
+        &mut self,
+        args: PayerSuppliedArgs,
+        payer: AccountId,
+        deposit: U128,
+        payment_token_decimals: u8,
+    ) -> Promise {
+        near_sdk::assert_self();
+
+        // Parse rate from oracle promise result
+        let rate = match env::promise_result(0) {
+            PromiseResult::NotReady => unreachable!(),
+            PromiseResult::Successful(value) => {
+                match serde_json::from_slice::<PriceEntry>(&value) {
+                    Ok(value) => value,
+                    Err(_e) => panic!("ERR_INVALID_ORACLE_RESPONSE"),
+                }
+            }
+            PromiseResult::Failed => panic!("ERR_FAILED_ORACLE_FETCH"),
+        };
+        // Check rate validity
+        assert!(
+            u64::from(args.max_rate_timespan) == 0
+                || rate.last_update >= env::block_timestamp() - u64::from(args.max_rate_timespan),
+            "Conversion rate too old (Last updated: {})",
+            rate.last_update,
+        );
+        let conversion_rate = u128::from(rate.price);
+        let decimals = u32::from(rate.decimals); // this is the conversion rate decimals, not the token decimals
+        let crypto_amount = Balance::from(args.amount)
+            * 10u128.pow(payment_token_decimals.into())
+            * 10u128.pow(decimals)
+            / conversion_rate
+            / ONE_FIAT;
+        let crypto_fee_amount = Balance::from(args.fee_amount)
+            * 10u128.pow(payment_token_decimals.into())
+            * 10u128.pow(decimals)
+            / conversion_rate
+            / ONE_FIAT;
+
+        let total_amount = crypto_amount + crypto_fee_amount;
```

Comment:
I don't think the `crypto_` prefix for `crypto_amount` and `crypto_fee_amount` actually add information.  I think they could be renamed to `amount` and `fee_amount`.

Same with `crypto_amount_args` and `crypto_fee_amount_args` below.